src/sentry/ratelimits/sliding_windows.py

```diff
@@ -0,0 +1,367 @@
+class Quota:
+    # The number of seconds to apply the limit to.
+    window_seconds: int
+
+    # A number between 1 and `window_seconds`. Since `window_seconds` is a
+    # sliding window, configure what the granularity of that window is.
+    #
+    # If this is equal to `window_seconds`, the quota resets to 0 every
+    # `window_seconds`.  If this is a very small number, the window slides
+    # "more smoothly" at the expense of having much more redis keys.
+    #
+    # The number of redis keys required to enforce a quota is `window_seconds /
+    # granularity_seconds`.
+    granularity_seconds: int
+
+    #: How many units are allowed within the given window.
+    limit: int
+
+    # Override the prefix given by RequestedQuota such that one can implement
+    # global limits + per-organization limits. The GrantedQuota will still only
+    # contain the prefix of the RequestedQuota
+    prefix_override: Optional[str] = None
+
+    def __post__init__(self) -> None:
+        assert self.window_seconds % self.granularity_seconds == 0
+
+    def iter_window(self, request_timestamp: int) -> Iterator[int]:
+        """
+        Iterate over the quota's window, yielding timestamps representing each granule.
+
+        This function is used to calculate keys for storing the number of
+        requests made in each granule.
+
+        The iteration is done in reverse-order (newest timestamp to oldest),
+        starting with the key to which a currently-processed request should be
+        added. That request's timestamp is `request_timestamp`.
+
+        * `request_timestamp / self.granularity_seconds`
+        * `request_timestamp / self.granularity_seconds - 1`
+        * `request_timestamp / self.granularity_seconds - 2`
+        * ...
+        """
+        value = request_timestamp // self.granularity_seconds
+
+        for granule_i in range(self.window_seconds // self.granularity_seconds):
+            value -= 1
+            assert value >= 0, value
+            yield value
+
+
+@dataclass(frozen=True)
+class RequestedQuota:
+    # A string that all redis state is prefixed with. For example
+    # `sentry-string-indexer:123` where 123 is an organization id.
+    #
+    # Note: You cannot control the redis sharding this way, so curly braces are
+    # forbidden.
+    prefix: str
+
+    # How much of each quota's limit is requested
+    requested: int
+
+    # Which quotas to check against. The requested amount must "fit" into all
+    # quotas.
+    quotas: Sequence[Quota]
+
+
+@dataclass(frozen=True)
+class GrantedQuota:
+    # The prefix from RequestedQuota
+    prefix: str
+
+    # How much of RequestedQuota.requested can actually be used.
+    granted: int
+
+
+class SlidingWindowRateLimiter(Service):
+    def __init__(self, **options: Any) -> None:
+        pass
+
+    def check_within_quotas(
+        self, requests: Sequence[RequestedQuota], timestamp: Optional[int] = None
+    ) -> Tuple[int, Sequence[GrantedQuota]]:
```

Comment:
can you document the return type? defining a `Timestamp = int` type will make this signature easier to understand